### PR TITLE
Added Advance to dilated wall clock.

### PR DESCRIPTION
**Pull request by hpidcock**
Adds `Advance(duration)` support to the dilated wall clock, allowing tests that need to
advance large chunks of time for the purpose of being fast, but unlike the testing clock
it allows tests to use a real wall clock.

This allows tests to stop being concerned with the implementation the test is testing,
but rather the behaviour.

### CODE REVIEW

File: testclock/dilated_test.go [hunk 1]
```diff
@@ -83,3 +84,101 @@ func (*dilatedClockSuite) TestFastNow(c *gc.C) {
...
+		case <-first:
+			n++
+		case <-second:
+			n++
+		case <-third:
+			n++
+		case <-fourth:
+			c.Fatal("timer that fired that should not have")
+		case <-done:
+			break out
```
Comment (wallyworld) [inline testclock/dilated_test.go:131]: This is a fallback but we can also break when n reaches 3 in any of the n++ lines above

File: testclock/dilated_test.go [hunk 2]
```diff
@@ -83,3 +84,101 @@ func (*dilatedClockSuite) TestFastNow(c *gc.C) {
...
+	ok = a.Stop()
+	c.Assert(ok, jc.IsFalse)
+	select {
+	case <-a.Chan():
+		c.Fatal("stopped clock fired")
+	case <-time.After(1 * time.Second):
+	}
+	time.Sleep(50 * time.Millisecond)
+	numGoAfter := runtime.NumGoroutine()
+	c.Assert(numGoAfter, gc.Equals, numGo, gc.Commentf("clock goroutine still running"))
```
Comment (wallyworld) [inline testclock/dilated_test.go]: This should be a retry loop with a short 50ms sleep between attempts.

File: testclock/dilated_test.go [hunk 3]
```diff
@@ -83,3 +84,101 @@ func (*dilatedClockSuite) TestFastNow(c *gc.C) {
...
+	numGo := runtime.NumGoroutine()
+	cl := testclock.NewDilatedWallClock(500 * time.Millisecond)
+	a := cl.NewTimer(1 * time.Second)
+	time.Sleep(250 * time.Millisecond)
+	ok := a.Reset(1 * time.Second)
+	c.Assert(ok, jc.IsTrue)
+	<-time.After(500 * time.Millisecond)
+	select {
+	case <-a.Chan():
+	case <-time.After(50 * time.Millisecond):
```
Comment (wallyworld) [inline testclock/dilated_test.go]: make this a shortWait const and use in all tests

File: testclock/dilated_test.go [hunk 4]
```diff
@@ -83,3 +84,101 @@ func (*dilatedClockSuite) TestFastNow(c *gc.C) {
...
+		c.Fatal("clock did not fire")
+	}
+	time.Sleep(50 * time.Millisecond)
+	numGoAfter := runtime.NumGoroutine()
+	c.Assert(numGoAfter, gc.Equals, numGo, gc.Commentf("clock goroutine still running"))
```
Comment (wallyworld) [inline testclock/dilated_test.go]: retry loop

File: testclock/dilated_test.go [hunk 5]
```diff
@@ -83,3 +84,101 @@ func (*dilatedClockSuite) TestFastNow(c *gc.C) {
...
```
Comment (wallyworld) [inline testclock/dilated_test.go]: ditto

File: testclock/dilated.go [hunk 6]
```diff
@@ -4,67 +4,215 @@
...
+		return <-stop
+	}
 }
 
-func (t dilatedWallTimer) Stop() bool {
-	return t.timer.Stop()
+func dilateTime(epoch, realNow time.Time,
+	realSecondDuration, dilatedOffset time.Duration) time.Time {
+	return epoch.Add(dilatedOffset).
+		Add(time.Duration(float64(realNow.Sub(epoch)) / realSecondDuration.Seconds()))
```
Comment (wallyworld) [inline testclock/dilated.go]: should we check that realNow > epoch

**Review by wallyworld — APPROVED** [review]

File: testclock/dilated_test.go (hunk 1 above)
Comment (hpidcock) [reply testclock/dilated_test.go:131]: This is waiting the duration to make sure the fourth timer doesn't fire.

File: testclock/dilated.go (hunk 6 above)
Comment (hpidcock) [reply testclock/dilated.go]: yeah probably should be
```
delta := realNow.Sub(epoch)
if delta < 0 {
    delta = time.Duration(0)
}
```

**Pull request merged**